fix: Handle HTTPExceptions from the TaskGroup

smithk86 opened the pull request:
<!--
By submitting this pull request, you agree to:
- follow [Litestar's Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)
- follow [Litestar's contribution guidelines](https://github.com/litestar-org/.github/blob/main/CONTRIBUTING.md)
- follow the [PSFs's Code of Conduct](https://www.python.org/psf/conduct/)
-->

### Pull Request Checklist

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [ ] Pre-Commit Checks were ran and passed
- [ ] Tests were ran and passed

### Description
- Moved the `get_exception_group()` helper from `tests/helpers.py` to `litestar/utils/helpers.py` and updated the test references
- Updated `litestar/_kwargs/kwargs_model.py` to use `_ExceptionGroup = get_exception_group()` and catch HTTPExceptions coming from the `TaskGroup` and throw the first one.
- The good: Easy implementation; at least one HTTPException at a time will have visibility
- The bad: If multiple HTTPExceptions are thrown in a single TaskGroup, only the first one will be seen.

### Close Issue(s)
- Fixes #2594

provinzkraut reviewed (changes requested):
Thanks @smithk86! I've left a few change requests.

smithk86 commented:
@provinzkraut I have reworked the changes with your suggestions. I also added an additional test to unit/test_dependency_batches.py to confirm `HTTPException` errors are handled correctly.

Cheers!

provinzkraut commented:
> @provinzkraut I have reworked the changes with your suggestions. I also added an additional test to unit/test_dependency_batches.py to confirm `HTTPException` errors are handled correctly.
> 
> Cheers!

Great! Can also also work in the type hunting stuff I opened a review comment about? As soon as that’s done this is good to go from my end.

smithk86 commented on litestar/_kwargs/kwargs_model.py:
```diff
@@ -45,6 +47,13 @@
     from litestar.dto import AbstractDTO
     from litestar.utils.signature import ParsedSignature
 
+_ExceptionGroup = get_exception_group()
```
@provinzkraut In the updated code, `_ExceptionGroup` is now referenced as an Exception in a try/except:

```python
try:
    async with create_task_group() as task_group:
        for dependency in batch:
            task_group.start_soon(resolve_dependency, dependency, connection, kwargs, cleanup_group)
except _ExceptionGroup as excgroup:
    raise excgroup.exceptions[0] from excgroup  # type: ignore[attr-defined]
```

provinzkraut replied:
Ah, sorry, totally missed that!

provinzkraut commented:
@JacobCoffee any idea what's going on with pdm there?

smithk86 commented:
> @JacobCoffee any idea what's going on in the CI with pdm?

I'm not sure if this helps at all but the checks were passing prior to Litestar v2.3.2 release. After the release, I rebased with main and did a force-push. The pdm stuff has been pretty sad ever since.

provinzkraut commented:
@smithk86 I merged the latest `main`; It has #2639, which should fix these errors.

provinzkraut commented:
@all-contributors add @smithk86 for code

provinzkraut commented:
@smithk86 Issues have been fixed and it's finally merged! Sorry it took so long! :)
